docs/rfcs/rfc-035-read-write-controls.md

```diff
@@ -0,0 +1,378 @@
+
+```gql
+type Movie @aggregations(enabled: false) {
+  title: String!
+}
+
+type Actor {
+  name: String!
+  movies: [Movie!]! @relationship(type: "ACTED_IN", direction: OUT) @aggregations 
+}
+
+extend type schema @aggregations
+```
+
+The above will:
+
+* Enable all root-level and nested aggregation fields
+  * However, the `Query.moviesAggregate` and `Actor.moviesAggregate` fields will be disabled by the `@aggregations` directive on the `Movie` type
+    * However, the `Actor.moviesAggregate` field will once again be enabled by the `@aggregations` directive on `Actor.movies`
+
+### Write operations
+
+#### Option 1: single `@write` directive
+
+This option takes an unopinionated view of root-level versus nested operations, encompassing them all in a single directive. These operations are:
+
+* `CREATE`
+* `UPDATE`
+* `DELETE`
+* `CONNECT`
+* `DISCONNECT`
+
+The directive for toggling these operations will be defined as:
+
+```gql
+enum WriteOperation {
+  CREATE
+  DELETE
+  UPDATE
+  CONNECT
+  DISCONNECT
+}
+
+directive @write(
+  operations: [WriteOperation!]! = [CREATE, UPDATE, DELETE, CONNECT, DISCONNECT]
+) on SCHEMA | OBJECT | FIELD_DEFINITION
+```
+
+To remove create operations from both the Mutation type and also nested operations, one can simply do:
+
+```gql
+extend schema @write(operations: [UPDATE, DELETE, CONNECT, DISCONNECT])
+```
+
+For a particular type, you can remove its create operation from the Mutation type and also from any nested relationship operations pointing _to_ it by doing:
+
+```gql
+type User @write(operations: [UPDATE, DELETE, CONNECT, DISCONNECT]) {
+  name: String!
+}
+```
+
+Finally, you can remove nested operations for a particular relationship:
+
+```gql
+type User {
+  name: String!
+}
+
+type Post {
+  author: User! @relationship(type: "HAS_AUTHOR", direction: OUT) @write(operations: [UPDATE, DELETE, CONNECT, DISCONNECT]) 
+}
+```
+
+It should be noted, that when used on a field definition, it is only valid when combined with a `@relationship` directive, and this should be confirmed during type definition validation.
+
+The `@write` directive can also be used in multiple locations to introduce complex rules:
+
+```gql
+type Comment @write(operations: [CREATE, DELETE, UPDATE, CONNECT, DISCONNECT]) {
+  content: String!
+}
+
+type Post {
+  comments: [Comment!]! @relationship(type: "HAS_COMMENT", direction: OUT) @write(operations: [CREATE])
+}
+
+extend schema @write(operations: [CREATE, UPDATE, CONNECT, DISCONNECT])
+```
+
+In the above example:
+
+* All delete operations are removed at the schema level
+  * The `deleteComments` Mutation field, however, is reintroduced by the `@write` directive on the `Comment` type
+* When performing a nested operation from the `Post` type, you can only create a `Comment`, and nothing else
+
+##### `connectOrCreate`
+
+The existence of the `connectOrCreate` operation will depend on both the `CREATE` and `CONNECT` operations being present in the `@write` directive.
```

Comment:
might help to think of the distinction between "what operations are allowed on the API?" vs "what GraphQl types are being generated?"

Some thoughts..

- "What operations are allowed on the API?"
  - would make sense to include `connectOrCreate` implicitly if both `connect` and `create` are allowed, otherwise not. 
  - would scale for any new operations we will introduce, as they will all be a composition of create/update/delete/connect/disconnect.
  - `write` and `read` make sense, as they have the feel of "operations"

- "What GraphQl types are being generated?"
  - would make sense to include `connectOrCreate` explicitly, independent of weather `connect` or `create` are include.
  - would mean that the operations will have to be updated with any new operations we will introduce.
  - in my opinion, `read` and `write` are a bit confusing terms, `include`/`exclude` (.. types) seem to reflect the meaning more.
  - instead of `readOnly`, `writeOnly` we would have something like `include(types: [READ])`, `include(types: [CREATE, UPDATE, DELETE])` or alternatively using `exclude` -> might consider adding a sort of  "sintactic sugar" option or all of these, like `WRITE:ALL` 
  - can get away with 2 directives instead of 4, available on schema, object_type, field_definition

Reply:
I see your reasoning @ale, I would say that the initial goal of this RFC was to substitute the current `@exclude` directive and provide some mechanism to limit the size of the schema generated, in that way I believe that this RFC answers the question "What GraphQl types are being generated?" and the Auth RFC answer the question "What operations are allowed on the API?", of course, this is just my interpretation!
On this point `would make sense to include connectOrCreate explicitly, independent of whether connect or create are include.`
I would say that this sentence is true only if the types used by the `connectOrCreate` operation are not shared between the `connect` and `create` types, that it may be true now, but it may be not be still true in the next versions of the library.
I 100% agree with this: `In my opinion, read and write are a bit confusing terms, include/exclude (.. types) seem to reflect the meaning more.`. 
The `@read/@write` directives give me the feeling that is not related to the schema generation phase but rather to some access control implemented in the resolvers!

Reply:
The whole principle of splitting into `@read` and `@write` was to reduce the complexity of `@exclude` - it quickly becomes a rabbit hole when you include read operations, `connectOrCreate`, future `updateOrCreate`, aggregations... You get the idea.

Having the directives split out like this means that a lot of the time, a user might be able to just use a directive without any arguments, versus less directives where you _always_ have to use arguments.

Personally, I would go for both `CONNECT` and `CREATE` being required for `connectOrCreate` - other than for reducing the size of the schema (in which removing 1 operation really won't make a difference), I don't see a use case where a user would want `connect` and `create`, but not `connectOrCreate`; you could use the two separate operations to the same effect.

Reply:
I do agree that not having to specify arguments on directives is elegant. However, except for `@read` the level of granularity is not that low, so chances are the user is going to specify arguments on the `@write` directive most of the time anyway.
Regarding `connect` + `create` vs `connectOrCreate`, the argument could be the other way around maybe - instead of having the types generated for both `connect` and `create`, a user may choose to only enable `connectOrCreate`.
The point of generated types vs operations is what feels awkward to me.

Reply:
To be fair, even if I agree that @read/@write does not reflect the idea as well as @include/@exclude, I also believe that this is easily solvable by good documentation.
Regarding the `connect` + `create`, I would with both being required as well. It's true that at this moment there is some specific type generated only for the `connectOrCreate` but this is just true only for the current design (I'm not implying that it will differ in the future), I believe that having a `CRUD` granularity will work independently by any design.

Reply:
Right, I have renamed `CONNECT` and `DISCONNECT` to `CREATE_RELATIONSHIP` and `DELETE_RELATIONSHIP`, respectively. This aligns with the nomenclature used in Subscriptions and also in the new authorization proposal. Additionally, it opens us up to controls over `UPDATE_RELATIONSHIP` in the future.

With this change, I feel requiring both `CREATE` and `CREATE_RELATIONSHIP` to use `connectOrCreate` makes sense, because we have abstracted away the nested operations from the directive.

Reply:
At the end of the day, you're talking very advanced users if they want only `connectOrCreate`, and there are other ways of removing fields from GraphQL schemas if desired.